common/cpp/src/flutter_screen_capture.cc

```diff
@@ -8,86 +8,213 @@
 
 namespace flutter_webrtc_plugin {
 
-void FlutterScreenCapture::EnumerateScreens(std::unique_ptr<MethodResult<EncodableValue>> result) {
-  SourceList sources;
+ FlutterScreenCapture::FlutterScreenCapture(FlutterWebRTCBase *base): base_(base) {
+std::string event_channel =
+      "FlutterWebRTC/desktopSourcesEvent";
+  event_channel_.reset(new EventChannel<EncodableValue>(
+    base_->messenger_, event_channel, &StandardMethodCodec::GetInstance()));
+
+  auto handler = std::make_unique<StreamHandlerFunctions<EncodableValue>>(
+		  [&](
+			  const flutter::EncodableValue* arguments,
+			  std::unique_ptr<flutter::EventSink<flutter::EncodableValue>>&& events)
+		  -> std::unique_ptr<StreamHandlerError<flutter::EncodableValue>> {
+	  event_sink_ = std::move(events);
+	  return nullptr;
+  },
+		  [&](const flutter::EncodableValue* arguments)
+	  -> std::unique_ptr<StreamHandlerError<flutter::EncodableValue>> {
+	  event_sink_ = nullptr;
+	  return nullptr;
+  });
+
+  event_channel_->SetStreamHandler(std::move(handler));
+ }
 
-  sources = base_->desktop_device_->EnumerateScreens();
-  
-  EncodableList list;
-  for (const auto& source : sources.std_vector()) {
-    // std::cout << " id: " << source.id.std_string() << " title: " << source.title.std_string() << " type: " << source.type << std::endl;
-    EncodableMap info;
-    info[EncodableValue("id")] = EncodableValue(source.id.std_string());
-    info[EncodableValue("name")] = EncodableValue(source.title.std_string());
-    info[EncodableValue("type")] = EncodableValue(source.type);
-    list.push_back(EncodableValue(info));
-  }
-
-  result->Success(EncodableValue(list));
-}
-
-void FlutterScreenCapture::EnumerateWindows(std::unique_ptr<MethodResult<EncodableValue>> result) {
-  SourceList sources;
-
-  sources = base_->desktop_device_->EnumerateWindows();
-  
-  EncodableList list;
-  for (const auto& source : sources.std_vector()) {
-    // std::cout << " id: " << source.id.std_string() << " title: " << source.title.std_string() << " type: " << source.type << std::endl;
-    EncodableMap info;
-    info[EncodableValue("id")] = EncodableValue(source.id.std_string());
-    info[EncodableValue("name")] = EncodableValue(source.title.std_string());
-    info[EncodableValue("type")] = EncodableValue(source.type);
-    list.push_back(EncodableValue(info));
+void FlutterScreenCapture::GetDesktopSources(const EncodableList &types, std::unique_ptr<MethodResult<EncodableValue>> result) {
+  EncodableList sources;
+  sources_.clear();
+  size_t size = types.size();
+  for (size_t i = 0; i < size; i++) {
+    std::string type = GetValue<std::string>(types[i]);
+    if (type == "window") {
+      scoped_refptr<RTCDesktopMediaList> window;
```

Comment:
Can we refactor this into a separate method? We seem to reuse this code elsewhere, and it's duplicated below for screen type (with the type seemingly the only difference).

Reply:
yup, it makes sense, I will refactor them to make the code more concise.